### ISSUE TITLE
New feature: choose a java main as well as a start class

### ISSUE THREAD

**Issue by dsyer** [spring-projects/spring-ide#138]
A Spring Boot app, when it runs in production, is usually launched by one of the Spring Boot jar launchers. It would be useful to be able to simulate that in the Boot dashboard in STS. E.g. I want to use the `PropertiesLauncher` or the `ThinJarLauncher` (from here: https://github.com/dsyer/spring-boot-thin-launcher). These classes would not be on the classpath of the app normally, so you would need to be able to select a jar that has them as well as the main class.

Those two launchers also have a lot of command line switches that we could expose to users in a more convenient way than typing them by hand. But that's a bonus.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

I suppose, in order to use these launchers, we'd actually have to package the code in the project before running it? Since these launchers are meant to be used from a packaged app? When we normally run the code from Eclipse, we run it directly from the stuff Eclipse compiler built, not by calling on the maven packaging process.

Anyhoo, I'm not entirely clear here what is expected. Some other questions in my mind are:

1. since the launcher's are not on the classpath, how/where do we find the appropriate launcher's to use? Are the choices baked into STS, do the versions of the launchers matter? Should the user be able to choose a specific version?

2. What kind of UI do we offer the user to choose a launcher (and/or version of the launcher).

3. How do we prepare/package the project to run with a particular launcher (do we try and run the maven/gradle build somehow? Or is there some other 'packaging utility' we can use? Or do we implement some kind of packages inside of STS itself? Or can we somehow run the launcher's directly with Eclipse project classpath?)

**Comment by kdvolder** [comment]
> so you would need to be able to select a jar that has them

I guess that is really the most crucial point. So the question, select the jar from where? Just a location on the user's hard-drive?

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

I did a little digging into how the 'thin jar launcher/wrapper' works. So in that case  the main type is called 'ThinJarWrapper'. To get a jar that can be placed on classpath so this type can be launched we need to do this:

1. Donwload or find this jar: /.m2/repository/org/springframework/boot/experimental/spring-boot-thin-layout/1.0.2.RELEASE/spring-boot-thin-layout-1.0.2.RELEASE.jar

2. Inside that jar pull-out the nested jar at  `/metainf/loader/spring-boot-thin-wrapper.jar`

This is the jar that actually contains the `ThinJarWrapper` class.  It seems rather unlikely a user will have this jar sitting by itself on his hard-drive. So a UI that makes the user point to a jar sitting on his filesystem probably isn't all that useful (as getting the jar will be a cumbersome process).

So... I can imagine that we implement something which is rather specific to the ThinJarLauncher and so STS would know exactly where to download the jar, extract the nested jar, and put that on the classpath when launching, but such a solution willnot  be very generally re-usable, it will *only* work for the ThinJarLauncher case.

Is that really what we are after here?

@davesyer I'm a bit at a loss to figure out what it is you really have in mind, so please clarify.

**Comment by dsyer** [comment]
The `ThinJarWrapper` can be downloaded in its own jar (you don't have to unpack it from another one): e.g. http://repo1.maven.org/maven2/org/springframework/boot/experimental/spring-boot-thin-wrapper/1.0.2.RELEASE/. Ditto the `PropertiesLauncher`. The best way for a user to locate these would be via maven co-ordinates.

I can simulate what I want STS to do here by simply adding a dependency to the POM and changing the Java main class in the launch configuration. So it doesn't seem really very complicated or specially crafted for the thin launcher. Just inconvenient and fiddly to do manually.

**Comment by kdvolder** [comment]
Okay, great. So basically a way to select jar via maven coordinates and specify a launcher class (maybe with some magics in the UI to limit the choices to main types that actually exist in the selected jar). You are right that doesn't sound too difficulty. 

I guess the part that seemed confusing to me, and so probably also other 'naive' users would be around knowing what jars they can actually use here, and where to find them, and the main classes in them.

Perhaps it would make most sense if we make this part easy for the user by curating a list of choices for them. E.g. there's some kind of pull-down with pre-populated list of 'launcher jars' they can choose from. Maybe we can also provide a 'type whatever you want' option for the experts. 

If we go with a curated list, the question is, who maintains that list, and how does STS find it. Do we just bake the list into STS? If so, how does STS team keep it up-to date when you guys make changes / fixes / new launchers.

I also have some potential confusion around the classpath given to the launcher. I presume in production, it would not have the actual jars from the project's dependencies itself on the classpath when launching the launcher, since, essentially, the launcher is the one responsible for setting up the classpath. So as such, I gather that *adding* the jar to your classpath isn't entirely faithfull to how it would work in production. Not sure whether this really matters or not. 

Finally I also wondered/worried about the 'packaging' since it seems the launcher stuff, like the thin launcher for example, consists a component to 'package' things and then also launcher thing that knows the packaging layout and uses that to setup a classpath for the boot app to run. It seems our 'simulation' here (picking a launcher jar + main type) only accounts for the launching stage and is implicitly assuming that pointing it at the eclipse-computed classpath instead of the 'real' packaging, is going to be okay for the launcher. If that's always true, regardless of the launcher/packager, then maybe I'm worried over nothing.

**Comment by dsyer** [comment]
> If we go with a curated list, the question is, who maintains that list, and how does STS find it.

We could stick it in spring.io or start.spring.io I guess, or you could scrape the versions from Maven Central. An expert option where the user can change co-ordinates would be good too.

> So as such, I gather that adding the jar to your classpath isn't entirely faithfull to how it would work in production. Not sure whether this really matters or not.

Correct. I think it probably does matter (so my simulation where I got it working isn't perfect), but we will only find out how much by doing it.

> implicitly assuming that pointing it at the eclipse-computed classpath instead of the 'real' packaging, is going to be okay for the launcher

Yeah, specifically we probably need to modify that classpath. At least to remove all the jar files. And let the user specify through system properties or command line arguments how to locate them again.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

I've thought about this more in 'the back of my head'. The more I think about it the more I think this is a bad idea. It feels too much like a 'half baked' idea with too much potential pitfalls and buggyness waiting to happen.

The biggest problem I have is that there doesn't seem to be a way to implement this correctly if all you can do is 'pick a starter class'. That's just not enough information to make it work correctly.

As I understand it, the starter class and the packaging (alogorithm / layout) are two sides of a coin. But without knowing what some 'random' launcher class expects, it seems impossible to ensure it will also get to see the layout it expects on its launched classpath. 

So... unless we can come up with a good idea to address this and somehow ensure that a launcher class, when launched, gets to see the packaging layout it expects to see, then I don't think we should implement this as a 'half baked' feature.

**Comment by kdvolder** [comment]
@dsyer Maybe you want to chime in? Perhaps it would help if we understand the typical use-case/motivation for this. 

It sounded to me that you wanted something that launched the app in a way that is really 'faithful' to how this would run in production. Is that right? If so then trying to do something that is 'somewhat similar' by mucking with the Eclipse provided classpath and compiled classes doesn't seem like a good idea. Then maybe we would be better of exploring some way to make it easier to build the 'production jar' with all its packaging and launcher support, and then launch that, directly from Eclipse.

In other words, for maven projects, call out to maven to build the jar, then launch that jar, and bypass all the usual Eclipse mechanics of creating a classpath with the stuff from the Eclipse workspace.

That sound like a good idea to you? If not, please help me understand what it is that you really want from this.

**Comment by dsyer** [comment]
Faithfuller is better, I guess. But you wouldn't want to lose debug support (e.g. stepping into library jars). If you can make that work through some eclipse wizardry then fantastic.

It's also worth noting that I phrased the original request in very general terms, but there are only a couple of actual examples of launchers. Maybe if we narrow it down it's easier to get started.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

Narrowing down will certainly help focus effort instead of drowning in lots of 'what if'. At least, if the goal is to somehow directly build support for these launcher into Eclipse/STS.

However, if your worry is that we would loose the debugger support if we just build and launch the jar by invoking an 'external tool'. I think that problem isn't too hard to solve. It just boils down to having the eclipse debugger properly configured by the Eclipse launcher so it can find the source jars, and passing the right JVM args to enable debug support in the launched process. So its a bit of plumbing, but I don't think there's anything insurmountable.

I also think that a option to do a 'production faithfull launch' (got to think of a better name for that in the ui still :-) makes more sense than asking the user to configure the thin (or whichever) launcher in the Eclipse ui to mirror what they have already put into their pom.

So... you think that some kind of checkbox to do a 'production faithful launch' as described, would fit your intended use-case(s)?

**Comment by dsyer** [comment]
I don't see how you will know what a "production faithful launch" means (even if you spell it correctly). You would have to parse the build configuration (pom.xml or build.gradle) and have some knowledge of what the spring boot plugin configuration meant. You might as well just ask the user to configure the launcher in an Eclipse dialog. Or maybe you have some other plan?

**Comment by kdvolder** [comment]
Configuring the launcher and then somehow executing it correctly from Eclipse seems really complicated to me. Maybe I just don't understand it well enough yet and its easier than I imagine. But, how do we get the 'packaging' right? Do I have to reimplement the packaging logic in terms of Eclipse api? What if the packaging changes in the future? What about different versions of the packaging / launcher? Do I have to implement all of them? 

Or do I have to find a way to somehow download them, add them to the IDEs classpath dynamically and execute them? That is a slippery slope, and having been down that kind of slope before with Grails IDE support... its a maintenance nightmare to try to deal with tolerating multiple versions of the internal launcher/packager APIs that may emerge over time as the launcher evolves.

So, it seems to me the really only way to be sure that the packaging is 'right' is to simply run a command like 'mvn clean package' or similar to execute whatever is defined by the user in the pom as 'how to package my app'. Then grab whatever jar comes out of that, and then launch that jar with something equivalent to 'java -jar <path-to-jar>'. If we do it that way, we don't need to know anything at all about how the jar is packaged and launched. As long as it produces something that is a runnable jar. So that's what I was thinking of doing.

This is what I mean with 'production faithful' though perhaps we need to come up with a better name for that.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Martin Lippert:) 

I discussed this with Dave and I think I know what we can/should do to support this use case in a few simple steps:

- we add a preference that allows users to specify the location of the thin-jar-launcher-wrapper on their local disc. Usually this will be in the local maven cache already, but we don't care that much about it and let the user instead identify this JAR for us.
- we add an option to the regular Spring Boot launch configs to choose between the `regular run mode` and the `thin jar run mode`. Default would be the `regular run mode` - that continues to do the same as we do at the moment, no changes involved.
- if the user switches that option to the `thin-jar-run-mode`, we change the classpath of the boot app to not include the maven dependencies and the compiled classes anymore, but instead the JAR of the thin jar launcher that the user had configured in the preferences.
- we also add two arguments to the launch as program arguments: `--thin.main=com.example.SampleApplication --thin.archive=target/classes` - the first one is the main class of the boot app, as already configured in the boot launch config, the second one points to the compiled classes (the entry that was on the regular classpath before without the maven dependencies)

As a second step we can automate and improve the way the thin-jar-launcher-wrapper-jar is being found/downloaded, but that is a different story for a potentially different release.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

Alright, will give it a try.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

>  add a preference that allows users to specify the location of the thin-jar-launcher-wrapper on their local disc. Usually this will be in the local maven cache already, but we don't care that much about it and let the user instead identify this JAR for us

I understand correctly, this would be a global preference? If so, then I would propose that , If the thin launcher pref has not been defined, we should, by default, hide the extra widgetry to use it from launch conf editor. Sound okay?

I would do it this way to avoid the extra clutter complexity of widgets that probably, most user won't really be using. And for those folks that know what they are doing, they can enable it by defining the preference.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Martin Lippert:) 

As a first step, this preference is global, yes.

It would be much better if the tooling would automatically resolve the correct thin jar launcher wrapper via Maven from the projects pom file, but that is a lot more complicated to realize and therefore postponed.

I am not a big fan of UI that hides widgets automatically in certain situations and not in others, but if you feel comfortable doing this, I am fine with that. As an alternative, I could also imagine that the dropdown box stays on the launch config, having the "regular launch" selected and nothing else if you don't have a thin-jar-launcher-wrapper configured. There could be a hyperlink next to the button to let the user jump to the preferences to configure the thin-jar-launcher-wrapper, this would be the usual way in Eclipse. But again, if the other option is easier to implement, feel free to do so.

As this is the first step into this feature area, we should keep it as simple as possible for now.

**Comment by dsyer** [comment]
+1 to "disabled but not hidden by default".

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

Just to check something...

 I am finding two kinds of thin jar launcher jars on maven central. 

http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22org.springframework.boot.experimental%22%20AND%20a%3A%22spring-boot-thin-launcher%22

There's a 'plain' .jar (spring-boot-thin-launcher-1.0.6.RELEASE.jar) and a 'exec' jar (spring-boot-thin-launcher-1.0.6.RELEASE-exec.jar).

I assume we are going to be using the 'exec' jar which includes a manifest.mf naming the main type?

**Comment by dsyer** [comment]
Wrong artifact. It's this one: http://repo1.maven.org/maven2/org/springframework/boot/experimental/spring-boot-thin-wrapper.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

I still think this is a bit of a 'half baked' feature. So that's why I'd 'hide' the widgets for now. If we get to a point where this works pretty well (not a buggy mess) and isn't really confusing (i.e. most people will actually know how to use that feature, or its 'self-explanatory') , then I think we can make it visible by default. I don't feel very confident right now that this first version will be bug-free, or something most people will actually know what to do with. So let's hide it for now, and consider it 'experimental'. That sound fair?

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

Implemented mechanic as desctribed. It works for maven project. For buildship project it doesn't work. Not really sure yet why, but I assume it is probably because Buildship project doesn't really produce everything the thin wrapper expecects in the project's output folder. 

The error I get is indicative of not having the dependencies on the classpath (.i.e class not found 'SptingBootApplication' ).

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

A quick inspection of the stuff in project output folder confirms my suspicion. For a buildship project there doesn't seem to be much else in there but the compiled classes. So really no metadata of any kind that would allow the launcher to determine the proper dependencies. That problem doesn't seem easily fixible to me.. (Unless we ask gradle to run the build process that produces a packaged jar and then launch that, as I sugested.

Going down the path that was sugested, using the output in the eclipse projects outputfolder, I think Ive reached the 'end of the line'. So we can either accept that it will only work for maven projects, or rethink how/what it is we want to implement here.

Unless someone has a good idea on how to make sure the thin launcher will see what it expects in a gradle project (is there some way to produce the correct sruff in the project output folder for gradle projects?

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

I've added some checking so that a error message shows up in the launch conf editor if 'Use thin launcher' option is selected on a non-maven project.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

My stuff is sitting in a branch at the moment. If it looks okay in CI build, I can merge it to master tomorrow and deliver the ticket.

**Comment by dsyer** [comment]
> Unless someone has a good idea on how to make sure the thin launcher will see what it expects in a gradle project (is there some way to produce the correct sruff in the project output folder for gradle projects?

Did you read the documentation https://github.com/dsyer/spring-boot-thin-launcher? There's a gradle plugin. Or you can use gradle native features to create a properties file or a pom.xml. I expect your launcher would work if you just include the plugin, as long as you set the archive root the right way.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Martin Lippert:) 

I disagree to the comment about the "half-baked" feature. I am sure we can think about a lot more automation or magic here, but the current design looks like a solid first step to me. Therefore I would prefer to disable the widgets over hide.

**Comment by kdvolder** [comment]
> Did you read the documentation https://github.com/dsyer/spring-boot-thin-launcher? There's a gradle plugin. Or you can use gradle native features to create a properties file or a pom.xml. I expect your launcher would work if you just include the plugin, as long as you set the archive root the right way.

I will have a look at that, but there are a number of difficulties I anticipate with that.

First:

> you can use gradle native features to create a properties file or a pom.xml. I expect your launcher would work if you just include the plugin, as long as you set the archive root the right way.

Those sound like instructions for a user to massage their project to make things work. So that will be painful for a user. If its not the user's job to prepare the project, then it would fall on the IDE, but its very hard for the IDE to parse and analyze gradle build files and make changes to them on a user's behalf, and we would be implementing such support in a 'vacuum' without any kind of framework to help with that task.

Second. Even if the project is 'prepared' like this, I seriously doubdt that BuildShip will execute the plugin code as part of a workspace project build. Though I'm not 100% sure of this, so its something I'll have to try.

Anyway, probably worth exploring this a bit for a possible future iteration of this feature.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

Created a ticket to explore Gradle support at a later time: https://www.pivotaltracker.com/story/show/151059713

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

Hmmm... I was just trying to verify whether it works in latest CI build prior delivering but there's something wrong. I get CNF exceptions for the thin launcher class. Reopening this ticket and investigating.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

I was able to reproduce in a runtime workbench, but then, I installed Eclipse SDKs (I needed these to get source code to debug) and then, when I tried again it actually worked.

My theory is that the problem is caused by changes in JDT launching infrastructure to support Java 9. 

It looks like the installation of the Eclipse SDK has somehow reverted some JDT bits back to a version of JDT without Java 9 support.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

I've got a fix, in a branch, but...

It needs to override a new method (getClasspathAndModulePath) added by Java 9 support. As this method doesn't exist in older versions of Eclipse API, I think it will wreak havoc in our builds running against the older Eclipse apis. 

I don't yet have a good idea how we can work around this. Can't really use reflection to override a method 'only if it exists'.

Andy's suggestion was that we somehow compile the code against newer api, but allow it to run against older api as well, and thinks that, in theory, that should work.

**Comment by spring-issuemaster** [comment]
(comment in Pivotal Tracker added by Kris De Volder:) 

Was able to use `MethodHandles` (stuff introduced in Java 7) to work around these compilation issues. MethodHandles,allows to 'reflectively' call super methods. So I used that to avoid having a direct compile dependency on the new method and still be able to override and call its super when it exists. Seems to work.
